File: backend/src/main/java/com/pgvaale/backend/controller/MaidController.java
package com.pgvaale.backend.controller;

import com.pgvaale.backend.entity.Maid;
import com.pgvaale.backend.entity.UserMaid;
import com.pgvaale.backend.entity.Feedback;
import com.pgvaale.backend.repository.MaidRepository;
import com.pgvaale.backend.repository.UserMaidRepository;
import com.pgvaale.backend.repository.FeedbackRepository;
import com.pgvaale.backend.security.JwtUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.web.bind.annotation.*;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;

@CrossOrigin(origins = "http://localhost:3000")
@RestController
@RequestMapping("/api/maid")
public class MaidController {
    
    @Autowired
    private MaidRepository maidRepository;
    
    @Autowired
    private UserMaidRepository userMaidRepository;
    
    @Autowired
    private FeedbackRepository feedbackRepository;
    
    @Autowired
    private JwtUtil jwtUtil;

    @GetMapping("/profile")
    public ResponseEntity<?> getProfile() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            System.out.println("Fetching profile for username: " + username);
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                System.out.println("Maid not found for username: " + username);
                return ResponseEntity.notFound().build();
            }
            
            Maid maid = maidOptional.get();
            System.out.println("Found maid: " + maid.getName() + ", Email: " + maid.getEmail());
            
            Map<String, Object> profile = new HashMap<>();
            profile.put("id", maid.getId());
            profile.put("name", maid.getName());
            profile.put("email", maid.getEmail());
            profile.put("phoneNumber", maid.getPhoneNumber());
            profile.put("region", maid.getRegion());
            profile.put("services", maid.getServices());
            profile.put("timing", maid.getTiming());
            profile.put("monthlySalary", maid.getMonthlySalary());
            profile.put("gender", maid.getGender());
            profile.put("approved", maid.isApproved());

            
            System.out.println("Profile data: " + profile);
            
            return ResponseEntity.ok(profile);
        } catch (Exception e) {
            System.out.println("Error fetching profile: " + e.getMessage());
            e.printStackTrace();
            return ResponseEntity.status(500).body("Error fetching profile: " + e.getMessage());
        }
    }

    @PostMapping("/profile")
    public ResponseEntity<?> updateProfile(@RequestBody Map<String, Object> profileData) {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Maid maid = maidOptional.get();
            
            // Update fields
            if (profileData.containsKey("name")) {
                maid.setName((String) profileData.get("name"));
            }
            if (profileData.containsKey("email")) {
                maid.setEmail((String) profileData.get("email"));
            }
            if (profileData.containsKey("phoneNumber")) {
                maid.setPhoneNumber((String) profileData.get("phoneNumber"));
            }
            if (profileData.containsKey("region")) {
                maid.setRegion((String) profileData.get("region"));
            }
            if (profileData.containsKey("services")) {
                maid.setServices((String) profileData.get("services"));
            }
            if (profileData.containsKey("timing")) {
                maid.setTiming((String) profileData.get("timing"));
            }
            if (profileData.containsKey("monthlySalary")) {
                Object salaryObj = profileData.get("monthlySalary");
                if (salaryObj instanceof Number) {
                    maid.setMonthlySalary(((Number) salaryObj).doubleValue());
                } else if (salaryObj instanceof String) {
                    try {
                        maid.setMonthlySalary(Double.parseDouble((String) salaryObj));
                    } catch (NumberFormatException e) {
                        return ResponseEntity.badRequest().body("Invalid salary format");
                    }
                }
            }

            
            Maid savedMaid = maidRepository.save(maid);
            return ResponseEntity.ok("Profile updated successfully");
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error updating profile: " + e.getMessage());
        }
    }

    @GetMapping("/dashboard")
    public ResponseEntity<?> getDashboardStats() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Maid maid = maidOptional.get();
            Long maidId = maid.getId();
            
            // Get counts
            Long pendingRequests = userMaidRepository.countByMaidIdAndStatus(maidId, UserMaid.RequestStatus.PENDING);
            Long acceptedJobs = userMaidRepository.countByMaidIdAndStatus(maidId, UserMaid.RequestStatus.ACCEPTED);
            
            // Calculate average rating
            Double averageRating = feedbackRepository.findAverageRatingByMaidId(maidId);
            
            // Get recent activity
            List<UserMaid> recentRequests = userMaidRepository.findAllRequestsByMaidId(maidId);
            
            Map<String, Object> dashboard = new HashMap<>();
            dashboard.put("maidName", maid.getName());
            dashboard.put("pendingRequests", pendingRequests);
            dashboard.put("acceptedJobs", acceptedJobs);
            dashboard.put("averageRating", averageRating != null ? averageRating : 0.0);
            dashboard.put("recentRequests", recentRequests);
            
            return ResponseEntity.ok(dashboard);
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error fetching dashboard: " + e.getMessage());
        }
    }

    @GetMapping("/requests")
    public ResponseEntity<?> getServiceRequests() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Long maidId = maidOptional.get().getId();
            List<UserMaid> requests = userMaidRepository.findAllRequestsByMaidId(maidId);
            
            return ResponseEntity.ok(requests);
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error fetching requests: " + e.getMessage());
        }
    }

    @PostMapping("/requests/{requestId}/status")
    public ResponseEntity<?> updateRequestStatus(
            @PathVariable Long requestId,
            @RequestBody Map<String, String> statusData) {
        try {
            String newStatus = statusData.get("status");
            if (newStatus == null) {
                return ResponseEntity.badRequest().body("Status is required");
            }
            
            Optional<UserMaid> requestOptional = userMaidRepository.findById(requestId);
            if (!requestOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            UserMaid request = requestOptional.get();
            
            // Verify the request belongs to the authenticated maid
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            
            if (!maidOptional.isPresent() || !request.getMaid().getId().equals(maidOptional.get().getId())) {
                return ResponseEntity.status(403).body("Unauthorized");
            }
            
            try {
                UserMaid.RequestStatus status = UserMaid.RequestStatus.valueOf(newStatus.toUpperCase());
                request.setStatus(status);
                
                // If status is ACCEPTED, update the accepted date time
                if (status == UserMaid.RequestStatus.ACCEPTED) {
                    request.setAcceptedDateTime(LocalDateTime.now());
                }
                
                userMaidRepository.save(request);
                return ResponseEntity.ok("Request status updated successfully");
            } catch (IllegalArgumentException e) {
                return ResponseEntity.badRequest().body("Invalid status");
            }
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error updating request status: " + e.getMessage());
        }
    }

    // Get requests by status
    @GetMapping("/requests/status/{status}")
    public ResponseEntity<?> getRequestsByStatus(@PathVariable String status) {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Long maidId = maidOptional.get().getId();
            
            try {
                UserMaid.RequestStatus requestStatus = UserMaid.RequestStatus.valueOf(status.toUpperCase());
                List<UserMaid> requests = userMaidRepository.findByMaidIdAndStatus(maidId, requestStatus);
                
                return ResponseEntity.ok(requests);
            } catch (IllegalArgumentException e) {
                return ResponseEntity.badRequest().body("Invalid status");
            }
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error fetching requests: " + e.getMessage());
        }
    }
    
    // Get all available maids (for users to hire)
    @GetMapping("/available")
    public ResponseEntity<?> getAvailableMaids(@RequestParam(required = false) String region) {
        try {
            List<Maid> maids;
            
            if (region != null && !region.trim().isEmpty()) {
                // Filter by region if provided
                maids = maidRepository.findByRegionAndApprovedTrue(region);
            } else {
                // Get all approved maids
                maids = maidRepository.findByApprovedTrue();
            }
            
            return ResponseEntity.ok(maids);
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error fetching available maids: " + e.getMessage());
        }
    }
    
    // Get maid details by ID
    @GetMapping("/{maidId}")
    public ResponseEntity<?> getMaidById(@PathVariable Long maidId) {
        try {
            Optional<Maid> maidOptional = maidRepository.findById(maidId);
            
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Maid maid = maidOptional.get();
            
            // Only return approved maids
            if (!maid.isApproved()) {
                return ResponseEntity.notFound().build();
            }
            
            return ResponseEntity.ok(maid);
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error fetching maid details: " + e.getMessage());
        }
    }

    // Delete maid account
    @DeleteMapping("/profile")
    public ResponseEntity<?> deleteAccount() {
        try {
            Authentication auth = SecurityContextHolder.getContext().getAuthentication();
            String username = auth.getName();
            
            Optional<Maid> maidOptional = maidRepository.findByUsername(username);
            if (!maidOptional.isPresent()) {
                return ResponseEntity.notFound().build();
            }
            
            Maid maid = maidOptional.get();
            Long maidId = maid.getId();
            
            // Delete all feedback for this maid
            List<Feedback> maidFeedback = feedbackRepository.findByMaidId(maidId);
            feedbackRepository.deleteAll(maidFeedback);
            
            // Delete all user-maid relationships for this maid
            List<UserMaid> userMaidRelations = userMaidRepository.findByMaidId(maidId);
            userMaidRepository.deleteAll(userMaidRelations);
            
            // Delete the maid account
            maidRepository.delete(maid);
            
            return ResponseEntity.ok("Account deleted successfully");
        } catch (Exception e) {
            return ResponseEntity.status(500).body("Error deleting account: " + e.getMessage());
        }
    }
} 

File: PGVaaleDotNetBackend/Controllers/MaidController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using PGVaaleDotNetBackend.Entities;
using PGVaaleDotNetBackend.Services;
using System.Security.Claims;

namespace PGVaaleDotNetBackend.Controllers
{
    [ApiController]
    [Route("api/maid")]
    [Authorize]
    public class MaidController : ControllerBase
    {
        private readonly MaidService _maidService;
        private readonly UserMaidService _userMaidService;

        public MaidController(MaidService maidService, UserMaidService userMaidService)
        {
            _maidService = maidService;
            _userMaidService = userMaidService;
        }

        [HttpGet("profile")]
        public async Task<IActionResult> GetProfile()
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                var profile = new
                {
                    id = maid.Id,
                    name = maid.Name,
                    email = maid.Email,
                    phoneNumber = maid.PhoneNumber,
                    region = maid.Region,
                    services = maid.Services,
                    timing = maid.Timing,
                    monthlySalary = maid.MonthlySalary,
                    gender = maid.Gender,
                    approved = maid.Approved
                };

                return Ok(profile);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching profile: {ex.Message}");
            }
        }

        [HttpPost("profile")]
        public async Task<IActionResult> UpdateProfile([FromBody] Dictionary<string, object> profileData)
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                // Update fields
                if (profileData.ContainsKey("name") && profileData["name"] != null)
                {
                    maid.Name = profileData["name"].ToString();
                }
                if (profileData.ContainsKey("email") && profileData["email"] != null)
                {
                    maid.Email = profileData["email"].ToString();
                }
                if (profileData.ContainsKey("phoneNumber") && profileData["phoneNumber"] != null)
                {
                    maid.PhoneNumber = profileData["phoneNumber"].ToString();
                }
                if (profileData.ContainsKey("region") && profileData["region"] != null)
                {
                    maid.Region = profileData["region"].ToString();
                }
                if (profileData.ContainsKey("services") && profileData["services"] != null)
                {
                    maid.Services = profileData["services"].ToString();
                }
                if (profileData.ContainsKey("timing") && profileData["timing"] != null)
                {
                    maid.Timing = profileData["timing"].ToString();
                }
                if (profileData.ContainsKey("monthlySalary"))
                {
                    if (profileData["monthlySalary"] is double salary)
                    {
                        maid.MonthlySalary = salary;
                    }
                    else if (profileData["monthlySalary"] is string salaryStr && double.TryParse(salaryStr, out double parsedSalary))
                    {
                        maid.MonthlySalary = parsedSalary;
                    }
                }

                await _maidService.SaveMaidAsync(maid);
                return Ok("Profile updated successfully");
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error updating profile: {ex.Message}");
            }
        }

        [HttpGet("dashboard")]
        public async Task<IActionResult> GetDashboardStats()
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                var maidId = maid.Id;

                // Get counts
                var pendingRequests = await _userMaidService.CountPendingRequestsByMaidIdAsync(maidId);
                var acceptedJobs = await _userMaidService.GetRequestsByMaidIdAsync(maidId);
                var acceptedCount = acceptedJobs.Count(um => um.Status == UserMaid.RequestStatus.ACCEPTED);

                // Get recent activity
                var recentRequests = await _userMaidService.GetRequestsByMaidIdAsync(maidId);

                var dashboard = new
                {
                    maidName = maid.Name,
                    pendingRequests = pendingRequests,
                    acceptedJobs = acceptedCount,
                    averageRating = 0.0, // TODO: Implement feedback system
                    recentRequests = recentRequests
                };

                return Ok(dashboard);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching dashboard: {ex.Message}");
            }
        }

        [HttpGet("requests")]
        public async Task<IActionResult> GetServiceRequests()
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                var requests = await _userMaidService.GetRequestsByMaidIdAsync(maid.Id);
                return Ok(requests);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching requests: {ex.Message}");
            }
        }

        [HttpPost("requests/{requestId}/status")]
        public async Task<IActionResult> UpdateRequestStatus(long requestId, [FromBody] Dictionary<string, string> statusData)
        {
            try
            {
                if (!statusData.ContainsKey("status"))
                {
                    return BadRequest("Status is required");
                }

                var newStatus = statusData["status"];
                if (!Enum.TryParse<UserMaid.RequestStatus>(newStatus.ToUpper(), out var status))
                {
                    return BadRequest("Invalid status");
                }

                // Verify the request belongs to the authenticated maid
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                var request = await _userMaidService.GetRequestByIdAsync(requestId);
                if (request == null)
                {
                    return NotFound("Request not found");
                }

                if (request.MaidId != maid.Id)
                {
                    return Forbid("Unauthorized");
                }

                await _userMaidService.UpdateRequestStatusAsync(requestId, status);
                return Ok("Request status updated successfully");
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error updating request status: {ex.Message}");
            }
        }

        [HttpGet("requests/status/{status}")]
        public async Task<IActionResult> GetRequestsByStatus(string status)
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                if (!Enum.TryParse<UserMaid.RequestStatus>(status.ToUpper(), out var requestStatus))
                {
                    return BadRequest("Invalid status");
                }

                var requests = await _userMaidService.GetRequestsByMaidIdAsync(maid.Id);
                var filteredRequests = requests.Where(r => r.Status == requestStatus).ToList();

                return Ok(filteredRequests);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching requests: {ex.Message}");
            }
        }

        [HttpGet("available")]
        [AllowAnonymous]
        public async Task<IActionResult> GetAvailableMaids([FromQuery] string? region)
        {
            try
            {
                List<Maid> maids;

                if (!string.IsNullOrEmpty(region))
                {
                    maids = await _maidService.GetMaidsByRegionAsync(region);
                }
                else
                {
                    maids = await _maidService.GetApprovedMaidsAsync();
                }

                return Ok(maids);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching available maids: {ex.Message}");
            }
        }

        [HttpGet("{maidId}")]
        [AllowAnonymous]
        public async Task<IActionResult> GetMaidById(long maidId)
        {
            try
            {
                var maid = await _maidService.GetMaidByIdAsync(maidId);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                if (!maid.Approved)
                {
                    return NotFound("Maid not available");
                }

                return Ok(maid);
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error fetching maid details: {ex.Message}");
            }
        }

        [HttpDelete("profile")]
        public async Task<IActionResult> DeleteAccount()
        {
            try
            {
                var username = User.FindFirst(ClaimTypes.Name)?.Value;
                if (string.IsNullOrEmpty(username))
                {
                    return Unauthorized("User not authenticated");
                }

                var maid = await _maidService.GetMaidByUsernameAsync(username);
                if (maid == null)
                {
                    return NotFound("Maid not found");
                }

                await _maidService.DeleteMaidAsync(maid.Id);
                return Ok("Account deleted successfully");
            }
            catch (Exception ex)
            {
                return StatusCode(500, $"Error deleting account: {ex.Message}");
            }
        }
    }
}